android/src/main/java/com/BV/LinearGradient/LinearGradientView.java

```diff
@@ -97,41 +97,133 @@ protected void onSizeChanged(int w, int h, int oldw, int oldh) {
         drawGradient();
     }
 
-    private float[] calculateGradientLocationWithAngle(float angle) {
-        float angleRad = (angle - 90.0f) * ((float)Math.PI / 180.0f);
-        float length = (float)Math.sqrt(2.0);
+    /**
+     * Gets the point of the element that a line perpendicular to the gradient line
+     * intersects first. This will always be a corner.
+     * @param angle the gradient line angle, in cartesian degrees
+     * @param size the size of the element
+     * @return the corner X and Y coordinates relative to the element center in cartesian
+     */
+    static private float[] getStartCornerToIntersect(float angle, int[] size) {
+        float halfWidth = size[0] / 2;
+        float halfHeight = size[1] / 2;
+        if (angle < 90f) {
+            // Bottom left
+            return new float[] { -halfWidth, -halfHeight };
+        } else if (angle < 180f) {
+            // Bottom right
+            return new float[] { halfWidth, -halfHeight };
+        } else if (angle < 270f) {
+            // Top right
+            return new float[] { halfWidth, halfHeight };
+        } else {
+            // Top left
+            return new float[] { -halfWidth, halfHeight };
+        }
+    }
 
-        return new float[]{
-                (float) Math.cos(angleRad) * length,
-                (float) Math.sin(angleRad) * length
-        };
+    /**
+     * Gets the start point assuming the angle is a multiple of 90 degrees
+     * @param angle the gradient line angle, in cartesian degrees
+     * @param size the size of the element
+     * @return the start point, relative to the angle center in cartesian
+     */
+    static private float[] getHorizontalOrVerticalStartPoint (float angle, int[] size) {
```

Comment:
If these 3 new methods should be static, can you please move them further up in the file? Right below the constructor on line 36/37 would be good.
The simple reason is because that's the default sorting order in Android Studio, and we want to avoid a major formatting-only commit that changes all these lines.
Also: There should be no space after the method name and `(`.